src/accessTokenProvider.js

```diff
@@ -0,0 +1,111 @@
+import {
+  isResponseUnauthorized,
+} from './services/http';
+
+export class AccessTokenProvider {
+  constructor(fetch, config) {
+    this.fetch = fetch;
+
+    this.config = config;
+    this.refreshAccessTokenPromise = null;
+    this.tokens = {
+      refreshToken: null,
+      accessToken: null,
+    };
+
+    this.isAuthorized = this.isAuthorized.bind(this);
+    this.refresh = this.refresh.bind(this);
+    this.clear = this.clear.bind(this);
+
+    this.resolveAccessToken = this.resolveAccessToken.bind(this);
+    this.fetchToken = this.fetchToken.bind(this);
+    this.handleFetchResolved = this.handleFetchResolved.bind(this);
+    this.handleTokenResolved = this.handleTokenResolved.bind(this);
+    this.handleError = this.handleError.bind(this);
+  }
+
+  /**
+   * Refreshes current access token with provided refresh token
+   */
+  refresh() {
+    // if token resolver is not authorized it should just resolve
+    if (!this.isAuthorized()) {
+      return Promise.resolve();
```

Comment:
Looking from outside, I would want for my requests to fail with 40x error with equal behavior as in case that I don't use fetch-token-intercept and didn't add proper Auth header. Also it would be useful to see additional error or warning from fetch-token-intercept about need to call `authorize` with instruction to pass `refreshToken`.